severn/api/constraint.py

```diff
@@ -0,0 +1,165 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+__all__ = ("Constraint",)
+
+import re
+from dataclasses import KW_ONLY, dataclass
+from functools import partial
+from typing import List, Optional, Tuple
+
+CONSTRAINT_PATTERN = re.compile(
+    r"""^
+    (?P<comparator>[<>~=!]{1,2})       # Comparator
+    (?:(?P<epoch>[1-9][0-9]*)!)?       # Epoch
+    (?P<release>[0-9]+(?:\.[0-9*]+)*)  # Major, minor, patch, etc.
+    (?:                                # Pre-release
+        (?P<pre_l>(?:a|b|rc))
+        (?P<pre_n>[0-9]+)?
+    )?
+    (?:.post(?P<post>[0-9]+))?         # Post-release
+    (?:.dev(?P<dev>[0-9]+))?           # Development release
+    $""",
+    re.VERBOSE | re.IGNORECASE,
+)
+MAX_VERSION = 0x3FFFFFFF
+PRE_LABEL_MAPPING = {"a": "alpha", "b": "beta"}
+
+
+def _eq_comparator(theirs: Tuple[int, ...], ours: Tuple[int, ...]) -> bool:
+    return theirs == ours
+
+
+def _ne_comparator(theirs: Tuple[int, ...], ours: Tuple[int, ...]) -> bool:
+    return theirs != ours
+
+
+def _gt_comparator(theirs: Tuple[int, ...], ours: Tuple[int, ...]) -> bool:
+    return theirs > ours
+
+
+def _ge_comparator(theirs: Tuple[int, ...], ours: Tuple[int, ...]) -> bool:
+    return theirs >= ours
+
+
+def _lt_comparator(theirs: Tuple[int, ...], ours: Tuple[int, ...]) -> bool:
+    return theirs < ours
+
+
+def _le_comparator(theirs: Tuple[int, ...], ours: Tuple[int, ...]) -> bool:
+    return theirs <= ours
+
+
+def _fuzzy_comparator(
+    specificity: int, theirs: Tuple[int, ...], ours: Tuple[int, ...]
+) -> bool:
+    return theirs >= ours and theirs[:specificity] == ours[:specificity]
+
+
+@dataclass()
+class Constraint:
+    comparator: str
+    release: List[int]
+    _: KW_ONLY
+    epoch: int = 0
+    alpha: int = MAX_VERSION
+    beta: int = MAX_VERSION
+    rc: int = MAX_VERSION
+    post: int = MAX_VERSION
+    dev: int = MAX_VERSION
+
+    def __post_init__(self) -> None:
+        comparator_mapping = {
+            "==": _eq_comparator,
+            "!=": _ne_comparator,
+            ">": _gt_comparator,
+            ">=": _ge_comparator,
+            "<": _lt_comparator,
+            "<=": _le_comparator,
+            "~=": partial(_fuzzy_comparator, len(self.release)),
+        }
+        self._cfunc = comparator_mapping[self.comparator]
+
+    @classmethod
+    def from_string(cls, raw: str, /) -> "Constraint":
+        if not (match := CONSTRAINT_PATTERN.match(raw)):
+            raise ValueError("constraint version must conform to PEP 440")
+
+        version = match.groupdict()
+        raw_release = version["release"]
+
+        if "*" in raw_release:
+            version["comparator"] = "~="
+            raw_release = raw_release.replace("*", "0")
+        release = [int(v) for v in raw_release.split(".")]
+
+        def safe_int(value: Optional[str], default: int = MAX_VERSION) -> int:
+            return int(value) if value is not None else default
```

Comment:
should this function be redefined every time we call from_string? or should we move it outside this method?

Reply:
Actually yeah probably.